atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampUtils.java

```diff
@@ -51,80 +55,102 @@
+        StringBuilder builder = new StringBuilder();
+        builder.append("BEGIN UNLOGGED BATCH\n"); // Safe, because all updates are on the same partition key
+
+        // Safe, because ordering does not apply in batches
+        checkAndSetRequest.entrySet().forEach(entry -> {
+            String columnName = entry.getKey();
+            byte[] expected = entry.getValue().getLhSide();
+            byte[] target = entry.getValue().getRhSide();
+            builder.append(constructCheckAndSetQuery(columnName, expected, target));
+        });
+        builder.append("APPLY BATCH;");
+        return toByteBuffer(builder.toString());
+    }
 
-        return ByteBuffer.wrap(PtBytes.toBytes(updateQuery));
+    private static String constructCheckAndSetQuery(String columnName, byte[] expected, byte[] target) {
+        if (expected == null) {
+            Preconditions.checkArgument(target != null, "[CAS] At least one of expected and target must be provided!");
+            return constructInsertIfNotExistsQuery(columnName, target);
+        }
+        if (target == null) {
+            // We treat this as a no op for now, which satisfies our purposes.
+            return "";
+        }
+        return constructUpdateIfEqualQuery(columnName, expected, target);
     }
 
     public static boolean wasOperationApplied(CqlResult cqlResult) {
-        Column appliedColumn = getNamedColumn(cqlResult, APPLIED_COLUMN);
+        Column appliedColumn = getNamedColumn(getColumnsFromOnlyRow(cqlResult), APPLIED_COLUMN);
         return Arrays.equals(appliedColumn.getValue(), SUCCESSFUL_OPERATION);
     }
 
-    public static long getLongValue(CqlResult cqlResult) {
-        Column valueColumn = getNamedColumn(cqlResult, VALUE_COLUMN);
-        if (Arrays.equals(INVALIDATED_VALUE, valueColumn.getValue())) {
-            throw new IllegalStateException("Couldn't extract a long from the invalidated value!");
-        }
-        return PtBytes.toLong(valueColumn.getValue());
+    public static long getLongValueFromApplicationResult(CqlResult result) {
+        return PtBytes.toLong(getNamedColumn(getColumnsFromOnlyRow(result), VALUE_COLUMN).getValue());
     }
 
-    private static Column getNamedColumn(CqlResult cqlResult, String columnName) {
-        List<Column> columns = getColumnsFromOnlyRow(cqlResult);
-        return Iterables.getOnlyElement(
-                columns.stream()
-                        .filter(column -> columnName.equals(PtBytes.toString(column.getName())))
-                        .collect(Collectors.toList()));
+    public static Map<String, byte[]> getValuesFromSelectionResult(CqlResult result) {
+        return result.getRows().stream()
+                .map(CqlRow::getColumns)
+                .collect(Collectors.toMap(
+                        cols -> PtBytes.toString(getNamedColumn(cols, COLUMN_NAME_COLUMN).getValue()),
```

Comment:
nit: the construction `getNamedColumn(x, y).getValue()` is used 3 times in the space of 10 lines. Could potentially be a helper method.